contrib/loaders/flash/fespi/riscv_fespi.c

```diff
@@ -0,0 +1,255 @@
+#define FESPI_INSN_ADDR_PROTO(x)  (((x) & 0x3) << 10)
+#define FESPI_INSN_DATA_PROTO(x)  (((x) & 0x3) << 12)
+#define FESPI_INSN_CMD_CODE(x)    (((x) & 0xff) << 16)
+#define FESPI_INSN_PAD_CODE(x)    (((x) & 0xff) << 24)
+
+/* Values */
+
+#define FESPI_CSMODE_AUTO         0
+#define FESPI_CSMODE_HOLD         2
+#define FESPI_CSMODE_OFF          3
+
+#define FESPI_DIR_RX              0
+#define FESPI_DIR_TX              1
+
+#define FESPI_PROTO_S             0
+#define FESPI_PROTO_D             1
+#define FESPI_PROTO_Q             2
+
+#define FESPI_ENDIAN_MSB          0
+#define FESPI_ENDIAN_LSB          1
+/* Timeout in target "loops" */
+#define FESPI_CMD_TIMEOUT         1000
+#define FESPI_PROBE_TIMEOUT       1000
+#define FESPI_MAX_TIMEOUT         30000
+
+enum {
+	ERROR_OK,
+	ERROR_FAIL
+};
+
+static int fespi_txwm_wait(volatile uint32_t *ctrl_base);
+static void fespi_disable_hw_mode(volatile uint32_t *ctrl_base);
+static void fespi_enable_hw_mode(volatile uint32_t *ctrl_base);
+static int fespi_wip(volatile uint32_t *ctrl_base);
+static int slow_fespi_write_buffer(volatile uint32_t *ctrl_base,
+		const uint8_t *buffer, unsigned offset, unsigned len);
+
+int main(volatile uint32_t *ctrl_base, uint32_t page_size,
+		const uint8_t *buffer, unsigned offset, uint32_t count)
+{
+	fespi_txwm_wait(ctrl_base);
+
+	/* Disable Hardware accesses*/
+	fespi_disable_hw_mode(ctrl_base);
+
+	/* poll WIP */
+	int retval = fespi_wip(ctrl_base);
+	if (retval != ERROR_OK)
+		goto err;
+
+	/* Assume page_size is a power of two so we don't need the modulus code. */
+	uint32_t page_offset = offset & (page_size - 1);
+
+	/* central part, aligned words */
+	while (count > 0) {
+		uint32_t cur_count;
+		/* clip block at page boundary */
+		if (page_offset + count > page_size)
+			cur_count = page_size - page_offset;
+		else
+			cur_count = count;
+
+		retval = slow_fespi_write_buffer(ctrl_base, buffer, offset, cur_count);
+		if (retval != ERROR_OK)
+			goto err;
+
+		page_offset = 0;
+		buffer += cur_count;
+		offset += cur_count;
+		count -= cur_count;
+	}
+
+err:
+	/* Switch to HW mode before return to prompt */
+	fespi_enable_hw_mode(ctrl_base);
+
+	return retval;
+}
+
+static uint32_t fespi_read_reg(volatile uint32_t *ctrl_base, unsigned address)
+{
+	return ctrl_base[address / 4];
+}
+
+static void fespi_write_reg(volatile uint32_t *ctrl_base, unsigned address, uint32_t value)
+{
+	ctrl_base[address / 4] = value;
+}
+
+static void fespi_disable_hw_mode(volatile uint32_t *ctrl_base)
+{
+	uint32_t fctrl = fespi_read_reg(ctrl_base, FESPI_REG_FCTRL);
+	fespi_write_reg(ctrl_base, FESPI_REG_FCTRL, fctrl & ~FESPI_FCTRL_EN);
+}
+
+static void fespi_enable_hw_mode(volatile uint32_t *ctrl_base)
+{
+	uint32_t fctrl = fespi_read_reg(ctrl_base, FESPI_REG_FCTRL);
+	fespi_write_reg(ctrl_base, FESPI_REG_FCTRL, fctrl | FESPI_FCTRL_EN);
+}
+
+static int fespi_txwm_wait(volatile uint32_t *ctrl_base)
+{
+	unsigned timeout = 1000;
```

Comment:
Just curious about the timeout value.  Is this derived from the specification of this type of flash device, or is it more of a pragmatic value that has been observed over time to be sufficient in practice?

Reply:
I think I carried over this number from the host-side code, where it reflects 1000ms. Here I don't have access to a clock, so we're just counting loops, which I'm sure are nowhere near 1ms per iteration. Empirically this works fine on HiFive1 and Arty.